library/kotlin/src/io/envoyproxy/envoymobile/EnvoyBuilder.kt

```diff
@@ -7,18 +7,25 @@ import io.envoyproxy.envoymobile.engine.EnvoyEngineImpl
 
 class ConfigurationException : Exception("Unresolved Template Key")
 
-class EnvoyBuilder internal constructor(
-    private val envoyConfiguration: EnvoyConfiguration
+open class EnvoyBuilder internal constructor(
+    private val envoyConfiguration: EnvoyConfiguration,
+    private var engineType: () -> EnvoyEngine
 ) {
   private var logLevel = LogLevel.INFO
   private var configYAML: String
-  private var engineType: () -> EnvoyEngine = { EnvoyEngineImpl() }
 
   private var connectTimeoutSeconds = 30
   private var dnsRefreshSeconds = 60
   private var statsFlushSeconds = 60
 
-  constructor() : this(EnvoyConfigurationImpl())
+  constructor() : this(EnvoyConfigurationImpl(), { EnvoyEngineImpl() })
+
+  /**
+   * Initialize with a specific implementation of `EnvoyEngine` to use for starting Envoy.
+   *
+   * A new instance of this engine will be created when `build()` is called.
+   */
+  internal constructor(engineType: () -> EnvoyEngine) : this(EnvoyConfigurationImpl(), engineType)
```

Comment:
Not that opinionated to undo

Reply:
Don't feel too strongly but one reason to revert would be for parity with iOS (since we have to do it the other way to prevent a segfault bug). Don't think adding a guarantee like this really provides much value since it's `internal`